### ISSUE TITLE
showerror(::MethodError) performance

### ISSUE THREAD

**Issue by green-nsk** [JuliaLang/julia#56740]
I ran into a problem where `showerror()` will take excessive time when there're a lot of candidates.

```Julia
julia> @time showerror(stdout, MethodError(getproperty, ()))
  ...
  0.007422 seconds (10.27 k allocations: 1012.766 KiB)

julia> using MyProtocol

julia> @time showerror(stdout, MethodError(getproperty, ()))
  ...
 22.118404 seconds (5.21 M allocations: 881.234 MiB, 1.06% gc time, 0.28% compilation time)
```
(there're about 37k overloads on `getproperty` in this example)

The problem is perhaps in [`show_method_candidates()`](https://github.com/JuliaLang/julia/blob/v1.10.7/base/errorshow.jl#L391) having some non-linear properties. I wish I could narrow it down further, but parts of it are quite arcane. The blunt solution would be only show method candidates when there'aren't too many candidate overloads.

For context, how come we have so many overloads. We generate protocol some protocol code, which makes message fields appear like properties. Goes something like this:

```Julia
struct MyMessageReader
    data::Ptr{Nothing}
end

@inline Base.getproperty(r::MyMessageReader, p::Symbol) = getproperty(r, Val(p))
@inline Base.getproperty(r::MyMessageReader, ::Val{:int1}) = unsafe_load(convert(Ptr{Int}, r.data))
```

It works fine mostly, but it's quite annoying when you make a typo in a property name, Julia throws `MethodError(getproperty, (MyMessageReader, Val{:my_prop1},))` and subsequent `showerror()` can take very long time. For now, we made `getproperty(r, ::Val)` overloads use module-local name, but it doesn't solve the problem completely because there're still quite a few overloads even within a single protocol (hundreds to thousands).

**Comment by green-nsk** [comment]
Further, looks like `methods()` eventually calls into `jl_matching_methods()` which further calls into `ml_matches()` which calls into [`sort_mlmatches()`](https://github.com/JuliaLang/julia/blob/9acf1129c91cddd9194f529ad9cc82afd2694190/src/gf.c#L4300). I assume this sorting from most specific to least specific is what causes non-linearity in time `methods()` takes.

Not sure if there's any takeaway from here. Perhaps, `showerror(::MethodError)` could be faster if there was a way to specify a cap on `methods()` output, but I'd be surprised if it was a priority for anyone.

**Comment by vtjnash** [comment]
I think the main takeaway is to generate less code (how many of those functions are meaningfully distinct and actually need to execute different code), and consider using if statements for branches instead of dispatch?

**Comment by Seelengrab** [comment]
I agree - if the code is generated anyway, why not use something like

```
function Base.getproperty(r::MyMessageReader, s::Symbol)
    if s === :int1
      # ...
    elseif s === :foo
        ...
    end
end
```

? I've had good success with this kind of pattern in the past, since the accessed property is a constant most of the time anyway.

**Comment by green-nsk** [comment]
Do you have any comments on constant propagation and type stability in branch-based dispatch?

I think the original design was done this way because it was perceived that we'd get more reliable type inference and type stability this way. 

I am not the author of the dispatch-based design, so there may have been other considerations too.

**Comment by vtjnash** [comment]
Closing since this is a package bug in protocol generation code
